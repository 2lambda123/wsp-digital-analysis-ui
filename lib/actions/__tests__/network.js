/* global describe, expect, it, jest */

describe('actions > network', () => {
  const nock = require('nock')
  const network = require('../network')

  it('decrementOutstandingRequests should work', () => {
    expect(network.decrementOutstandingRequests()).toMatchSnapshot()
  })

  it('incrementOutstandingRequests should work', () => {
    expect(network.incrementOutstandingRequests()).toMatchSnapshot()
  })

  describe('> serverAction', () => {
    it('should process successful request', async () => {
      nock('http://mockHost.com/')
        .get('/mockPath')
        .reply(200, 'success')

      const nextFn = jest.fn()
      const onErrorFn = jest.fn()
      const actionResult = network.serverAction({
        next: nextFn,
        onError: onErrorFn,
        options: {},
        url: '/mockPath'
      })

      // expect incrementOutstandingRequests action
      expect(actionResult[0]).toMatchSnapshot()

      // get response
      const response = await actionResult[1]

      // expect to receive array of actions
      // - decrementOutstandingRequests
      // - call to nextFn
      expect(response[0]).toMatchSnapshot()
      expect(nextFn).toHaveBeenCalled()

      // onError should not have been called
      expect(onErrorFn).not.toHaveBeenCalled()
    })

    it('should handle unsuccessful request when onError is provided', async () => {
      nock('http://mockHost.com/')
        .get('/mockPath')
        .reply(500, 'error')

      const nextFn = jest.fn()
      const onErrorFn = jest.fn()
      const actionResult = network.serverAction({
        next: nextFn,
        onError: onErrorFn,
        options: {},
        url: '/mockPath'
      })

      // expect incrementOutstandingRequests action
      expect(actionResult[0]).toMatchSnapshot()

      // get response
      const response = await actionResult[1]

      // expect to receive array of actions
      // - decrementOutstandingRequests
      // - call to onErrorFn
      expect(response[0]).toMatchSnapshot()
      expect(onErrorFn).toHaveBeenCalled()

      // next should not have been called
      expect(nextFn).not.toHaveBeenCalled()
    })

    it('should handle unsuccessful request when onError is not provided', async () => {
      nock('http://mockHost.com/')
        .get('/mockPath')
        .reply(500, 'error')

      const nextFn = jest.fn()
      const actionResult = network.serverAction({
        next: nextFn,
        options: {},
        url: '/mockPath'
      })

      // expect incrementOutstandingRequests action
      expect(actionResult[0]).toMatchSnapshot()

      // get response
      const response = await actionResult[1]

      // expect to receive array of actions
      // - decrementOutstandingRequests
      // - lockUiWithError
      expect(response).toMatchSnapshot()

      // next should not have been called
      expect(nextFn).not.toHaveBeenCalled()
    })
  })
})
