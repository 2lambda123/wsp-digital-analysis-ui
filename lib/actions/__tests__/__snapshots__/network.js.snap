exports[`actions > network > serverAction should handle unsuccessful request when onError is not provided 1`] = `
Object {
  "type": "increment outstanding requests",
}
`;

exports[`actions > network > serverAction should handle unsuccessful request when onError is not provided 2`] = `
Array [
  Object {
    "type": "decrement outstanding requests",
  },
  Object {
    "error": true,
    "payload": Object {
      "detailMessage": "Internal Server Error",
      "error": "Network error. Check your connection and reload the page?",
    },
    "type": "lock ui with error",
  },
]
`;

exports[`actions > network > serverAction should handle unsuccessful request when onError is provided 1`] = `
Object {
  "type": "increment outstanding requests",
}
`;

exports[`actions > network > serverAction should handle unsuccessful request when onError is provided 2`] = `
Object {
  "type": "decrement outstanding requests",
}
`;

exports[`actions > network > serverAction should process successful request 1`] = `
Object {
  "type": "increment outstanding requests",
}
`;

exports[`actions > network > serverAction should process successful request 2`] = `
Object {
  "type": "decrement outstanding requests",
}
`;

exports[`actions > network decrementOutstandingRequests should work 1`] = `
Object {
  "type": "decrement outstanding requests",
}
`;

exports[`actions > network incrementOutstandingRequests should work 1`] = `
Object {
  "type": "increment outstanding requests",
}
`;
